### PR TITLE
Allow other applications to hook into `showerror`

### LINKED ISSUE #68: Remove LSP specific error codes
https://github.com/julia-vscode/JSONRPC.jl/blob/5adc04bf4df6a220dda20896a8ecc1fd396a9346/src/core.jl#L18-L30

These are not listed in the [JSON-RPC](https://www.jsonrpc.org/specification) specification under section 5.1, so should probably be removed, as they are in the application defined section and may be different for other servers/clients.

A potential fix would be to have a wrapper error in LanguageServer.jl for the (then missing) custom printing, which would be thrown in places where JSON-RPC errors can occur by catching & rethrowing with the wrapper type. If that is acceptable, I can try to make PRs here and at LanguageServer.jl to implement this.

### PR DESCRIPTION
**Pull request by Seelengrab**
Closes #68.

This change allows other applications to customize `showerror(::IO, ::JSONRPCError)` depending on their specific needs and chosen meanings for their errors. It also introduces the fixed error codes as constants, to allow them to be transparently referenced in other places of the codebase.

Regarding `serverErrorStart` and `serverErrorEnd` - I don't know exactly how they're used in the VSCode extension/LS.jl (are the messages themselves matched anywhere?), so these could be added explicitly until those dependents are explicitly adding them to `RPCErrorStrings`. I've added them for now to make tests pass, but having them implicit seems cleaner to me.

For every PR, please check the following:
- [x] End-user documentation check. If this PR requires end-user documentation in the Julia VS Code extension docs, please add that at https://github.com/julia-vscode/docs.
- [x] Changelog mention. If this PR should be mentioned in the CHANGELOG for the Julia VS Code extension, please open a PR against https://github.com/julia-vscode/julia-vscode/blob/master/CHANGELOG.md with those changes.

### CODE REVIEW

**Comment by Seelengrab** [comment]
Not quite sure what the coverage diff is about - it seems to be inconsistent in terms of what it compares? Or I'm clicking on the wrong thing, not sure.

File: src/core.jl [hunk 1]
```diff
@@ -1,35 +1,86 @@
...
+    INVALID_PARAMS => "InvalidParams",
+    INTERNAL_ERROR => "InternalError",
+    [ i => "ServerError" for i in SERVER_ERROR_START:SERVER_ERROR_END]...,
+    # TODO: these should be removed - they are in the application/implementation specific range
+    -32000 => "serverErrorEnd",
+    -32099 => "serverErrorStart",
+    -32002 => "ServerNotInitialized",
+    -32001 => "UnknownErrorCode",
+    -32800 => "RequestCancelled",
+	-32801 => "ContentModified"
```
Comment (pfitzseb) [inline src/core.jl:79]: Not sure what you mean by that. These are implementation specific -- this in fact is the implementation they are specified by.

Comment (Seelengrab) [reply src/core.jl:79]: Yeah, I'll have to update these a bit - the first two (`serverErrorEnd`/`serverErrorStart`) at least aren't error codes at all, so shouldn't have an error message associated with them by default. The next two are definitely ok, but the last two are (I think) LSP specific, and not necessarily JSONRPC; there isn't necessarily support for cancellable requests or content that would have been modified mid-request (in particular if the underlying implementation is idempotent) by a codebase using this. They're also outside the range of reserved error codes for JSON-RPC servers, so IMO should not live here.

**Comment by Seelengrab** [comment]
The original intention here was to keep the LSP specific ones around until the time that LanguageServer.jl can/wants to integrate the functionality, but since that would require compat bumping sooner or later anyway, might as well just remove them now and do a lockstep release with appropriate compat upper bounds and/or upper bounding previous versions in the registry.

The question is just - is this a technically breaking release, since it changes the error message of LanguageServer.jl from e.g. `ContentModified` to `Unknown`?

File: src/core.jl [hunk 2]
```diff
@@ -1,35 +1,99 @@
...
+"""
+    SERVER_ERROR_END
+
+The end of the range of server-reserved errors.
+
+These are JSON-RPC server errors that are free for the taking
+for JSON-RPC server implementations. Applications making use of
+this library should NOT define new errors in this range.
+"""
+const SERVER_ERROR_END       = -32000
```
Comment (pfitzseb) [inline src/core.jl:28]: Nitpick, but can you fix the whitespace here and below pls?

Comment (Seelengrab) [reply src/core.jl:28]: Sure, that was just leftover from when they were defined together in a bigger list.

**Comment by Seelengrab** [comment]
> Mh, not sure. The error type doesn't change after all. I don't mind tagging a 2.0 for correctness either way, but imho that wouldn't really be necessary.

Yeah, I don't mind either way :) Whichever is more convenient to have this percolate through the dependencies without causing too much disruption.

Is there some NEWS file or release notes specific to JSONRPC.jl that this should be mentioned in as a new feature?

**Review by pfitzseb — APPROVED** [review]

**Pull request merged**